.github/workflows/build_and_test.yaml

```diff
@@ -0,0 +1,162 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+name: Build and Run Tests
+
+on:
+  push:
+    # Publish only on `master`
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+  release:
+    types: [published, edited]
+
+jobs:
+  build:
+    name: Build repository
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v2
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      # Stores external dependencies, can be further improved with Gradle 6.1
+      - name: Cache Gradle Dependencies
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          # Only rebuild cache if build.gradle is changed
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle') }}
+          restore-keys: ${{ runner.os }}-gradle
+      - name: Build repository
+        run: |
+          ./gradlew --no-daemon clean build -x test -x javadoc -x findbugsMain -x findbugsTest -x checkstyleMain -x checkstyleJmh -x checkstyleTest -Dorg.gradle.parallel=true
+
+  test_coverage:
```

Comment:
Other than run_tests, should other job also need to specify `needs:build` ?

Reply:
So interestingly enough none of the builds need to be dependent on each other, because gradle should be building the files needed if it hasn't been done so yet on the runner. But I can make the others dependent on `build` too, since I think it would make more logical sense

Reply:
If every job checks out a clean repo, then the project will get rebuilt every time. This can increase the time of CI pipeline, as things will get rebuilt 3 times. This can also introduce surprising bugs. For example, if we use a wildcard dependency, the build job can pull one version, and test job can pull a newer one.

Reply:
Yup that's a good point, I made the other jobs dependent on the first build job so now it should only download dependencies once. The dependencies should be stored by the gradle cache after `build` which is shared across jobs, but if it's wildcard then gradle might force a dependency download/check. I think in this scenario if it causes a problem it's acceptable as the later steps would have the newer dependency, and if it causes the tests to break then we should really be building against an exact version.